File: Leetcode/dfs_recursion_backtracking/133_CloneGraph/Solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;

    public Node() {}

    public Node(int _val,List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
class Solution {
    HashMap<Integer, Node> dic = new HashMap<>();
    public Node cloneGraph(Node node) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        Set<Integer> visited = new HashSet<>();
        dfs(node, visited, graph);
        for(int k: graph.keySet()){
            dic.put( k, new Node(k, new LinkedList<Node>()) );
        }
        for(int k: graph.keySet()){
            for(int v: graph.get(k))
                dic.get(k).neighbors.add(dic.get(v));
        }
        return dic.get(node.val);
    }
        private void dfs(Node node, Set<Integer> visited,
                         Map<Integer, List<Integer>> graph){
            if(visited.contains(node.val))
                return;
            visited.add(node.val);
            graph.put(node.val, new LinkedList<Integer>());
            for(Node n: node.neighbors){
                graph.get(node.val).add(n.val);
                dfs(n, visited, graph);
            }
        }
}

File: Leetcode/dfs_recursion_backtracking/133_CloneGraph/Solution.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, neighbors):
        self.val = val
        self.neighbors = neighbors
"""
class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        graph = collections.defaultdict(list)
        visited = set()
        def dfs(node):
            if node.val in visited:
                return None
            visited.add(node.val)
            graph[node.val] = []
            for n in node.neighbors:
                graph[node.val].append(n.val)
                dfs(n)
        dfs(node)
        dic = {}
        for k in graph:
            dic[k] = Node(k, [])
        for k in graph:
            for v in graph[k]:
                dic[k].neighbors.append(dic[v])
        return dic[node.val]
